### PR TITLE
Make `Style/RedundantFreeze` aware of methods that will produce frozen objects

### PR DESCRIPTION
**Pull request by rrosenblum**
This was extract out of the PR to add strict mode to `Style/MutableConstants`, #6126. This isn't an exhaustive list of methods that return immutable values, but it is at least a start.

### CODE REVIEW

File: config/default.yml [hunk 1]
```diff
@@ -3867,6 +3867,7 @@ Style/RedundantFreeze:
   Description: "Checks usages of Object#freeze on immutable objects."
   Enabled: true
   VersionAdded: '0.34'
+  VersionAdded: '0.66'
```
Comment (mikegee) [inline config/default.yml]: Is it ok to have two `VersionAdded` or was this one supposed to be `VersionChanged`?

Comment (rrosenblum) [reply config/default.yml]: It's supposed to be `VersionChanged `. Copy/paste error.

File: CHANGELOG.md [hunk 2]
```diff
@@ -48,6 +52,7 @@
 
 * [#6704](https://github.com/rubocop-hq/rubocop/pull/6704): Add new `Rails/ReflectionClassName` cop. ([@Bhacaz][])
 * [#6643](https://github.com/rubocop-hq/rubocop/pull/6643): Support `AllowParenthesesInCamelCaseMethod` option on `Style/MethodCallWithArgsParentheses` `omit_parentheses`. ([@dazuma][])
+* Add an experimental strict mode to `Style/MutableConstant` that will freeze all constants, rather than just literals. ([@rrosenblum][])
```
Comment (bbatsov) [inline CHANGELOG.md]: I think you've added this here accidentally.

File: CHANGELOG.md [hunk 3]
```diff
@@ -22,6 +22,10 @@
 * [#6806](https://github.com/rubocop-hq/rubocop/pull/6806): Remove `powerpack` dependency. ([@dduugg][])
 * [#6810](https://github.com/rubocop-hq/rubocop/pull/6810): Exclude gemspec file by default for `Metrics/BlockLength` cop. ([@koic][])
 
+### New features
```
Comment (bbatsov) [inline CHANGELOG.md]: This should be moved to changes probably.

File: spec/rubocop/cop/style/redundant_freeze_spec.rb [hunk 4]
```diff
@@ -23,6 +23,11 @@
   it_behaves_like 'immutable objects', '1.5'
   it_behaves_like 'immutable objects', ':sym'
   it_behaves_like 'immutable objects', ':""'
+  it_behaves_like 'immutable objects', "ENV['foo']"
+  it_behaves_like 'immutable objects', "'foo'.count"
+  it_behaves_like 'immutable objects', '(1 + 2)'
```
Comment (bbatsov) [inline spec/rubocop/cop/style/redundant_freeze_spec.rb]: I guess we also need tests which show what happens for objects other than numbers.

Comment (rrosenblum) [reply spec/rubocop/cop/style/redundant_freeze_spec.rb]: I added tests for comparing strings, comparing variables, adding string, and adding variables.

**Pull request merged**